### PR TITLE
Update Serde to 1.0

### PR DESCRIPTION
**Pull request by mike-engel**
This is PR updates serde to 1.0. The README and Cargo file have been updated to use 1.0. This will close out #27.

### CODE REVIEW

**Comment by mike-engel** [comment]
As of right now, there are two lifetime issues [here](https://github.com/mike-engel/rust-jwt/blob/serde-1.0/src/serialization.rs#L31) and [here](https://github.com/mike-engel/rust-jwt/blob/serde-1.0/src/serialization.rs#L41). I've tried using `str::from_utf8`, but that's no good either since then `decoded` goes out of scope too quick.

@Keats I'll keep trying here, but I don't think I have enough Rust experience with how to combat these two lifetime issues. If you have any ideas, let me know!

```
error: `s` does not live long enough
  --> src/serialization.rs:31:18
   |
31 |     Ok(from_str(&s)?)
   |                  ^ does not live long enough
32 | }
   | - borrowed value only lives until here
   |
note: borrowed value must be valid for the lifetime 'a as defined on the body at 27:85...
  --> src/serialization.rs:27:86
   |
27 |   pub fn from_jwt_part<'a, B: AsRef<str>, T: Deserialize<'a>>(encoded: B) -> Result<T> {
   |  ______________________________________________________________________________________^ starting here...
28 | |     let decoded = base64::decode_config(encoded.as_ref(), base64::URL_SAFE_NO_PAD)?;
29 | |     let s = String::from_utf8(decoded)?;
30 | |
31 | |     Ok(from_str(&s)?)
32 | | }
   | |_^ ...ending here

error: `s` does not live long enough
  --> src/serialization.rs:41:31
   |
41 |     let claims: T = from_str(&s)?;
   |                               ^ does not live long enough
...
44 | }
   | - borrowed value only lives until here
   |
note: borrowed value must be valid for the lifetime 'a as defined on the body at 37:40...
  --> src/serialization.rs:37:41
   |
37 |        -> Result<(T, Map<String, Value>)> {
   |  _________________________________________^ starting here...
38 | |     let decoded = base64::decode_config(encoded.as_ref(), base64::URL_SAFE_NO_PAD)?;
39 | |     let s = String::from_utf8(decoded)?;
40 | |
41 | |     let claims: T = from_str(&s)?;
42 | |     let map: Map<_, _> = from_str(&s)?;
43 | |     Ok((claims, map))
44 | | }
   | |_^ ...ending here
```

**Comment by Keats** [comment]
Thanks!
Actually I just noticed there is a https://docs.serde.rs/serde/de/trait.DeserializeOwned.html trait that might be more what we want since as a library we don't really know how the end struct is going to be used.

Is that how it's meant to be used @dtolnay (sorry for the random ping)?

**Comment by dtolnay** [comment]
I always appreciate a ping! Yes, both `from_jwt_part` and `from_jwt_part_claims` require `T: DeserializeOwned` because they are deserializing from the base64-decoded data so they shouldn't be allowed to borrow anything (the base64-decoded goes out of scope when the function returns).

I will leave a few review comments as well.

File: src/lib.rs [hunk 1]
```diff
@@ -101,7 +97,13 @@ macro_rules! expect_two {
 /// // Claims is a struct that implements Deserialize
 /// let token_data = decode::<Claims>(&token, "secret", Algorithm::HS256, &Validation::default());
 /// ```
-pub fn decode<T: Deserialize>(token: &str, key: &[u8], algorithm: Algorithm, validation: &Validation) -> Result<TokenData<T>> {
+pub fn decode<T>(token: &str,
+                 key: &[u8],
+                 algorithm: Algorithm,
+                 validation: &Validation)
+                 -> Result<TokenData<T>>
+    where for<'a> T: Deserialize<'a>
```
Comment (dtolnay) [inline src/lib.rs]: This would be more readable as `T: DeserializeOwned`. This function requires that `T` can deserialize without borrowing any data because the string it is deserialized from is thrown away before this function returns.

File: src/serialization.rs [hunk 2]
```diff
@@ -4,15 +4,17 @@ use serde::ser::Serialize;
...
-use errors::{Result};
+use errors::Result;
 use header::Header;
 
 
 /// The return type of a successful call to decode
 #[derive(Debug)]
-pub struct TokenData<T: Deserialize> {
+pub struct TokenData<T>
+    where for<'a> T: Deserialize<'a>
```
Comment (dtolnay) [inline src/serialization.rs]: This bound should not be here. Structs in Rust should never have trait bounds. Trait bounds are for behavior.

```rust
impl<T> Data<T> where T: Behavior { /* behavior, okay */ }

fn behave<T>() where T: Behavior { /* also behavior, also okay */ }

struct Data<T> where T: Behavior { /* don't do this */ }
```

The only exception is things like [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) that use associated types to define data.

File: src/crypto.rs [hunk 3]
```diff
@@ -29,10 +29,8 @@ pub enum Algorithm {
 /// The actual HS signing + encoding
 fn sign_hmac(alg: &'static digest::Algorithm, key: &[u8], signing_input: &str) -> Result<String> {
     let signing_key = hmac::SigningKey::new(alg, key);
-    Ok(base64::encode_config(
-        hmac::sign(&signing_key, signing_input.as_bytes()).as_ref(),
-        base64::URL_SAFE_NO_PAD
-    ))
+    Ok(base64::encode_config(hmac::sign(&signing_key, signing_input.as_bytes()).as_ref(),
```
Comment (dtolnay) [inline src/crypto.rs]: Next time please make whitespace changes in a separate commit. They makes it hard to find and review the thing you are actually changing.

Comment (Keats) [inline src/crypto.rs]: Agreed on that, @mike-engel can you revert all the formatting changes for now?
Is it rustfmt using the latest config? I thought it would be more block indented than that

Comment (dtolnay) [inline src/crypto.rs]: Block style is currently recommended by the fmt-rfc folks but not the default in rustfmt. [This is the settings they recommend.](https://github.com/serde-rs/serde/blob/v1.0.0/rustfmt.toml)

Comment (dtolnay) [inline src/crypto.rs]: More official source I guess: [here](https://github.com/rust-lang-nursery/rustfmt/blob/master/rfc-rustfmt.toml).

Comment (mike-engel) [inline src/crypto.rs]: Yeah, sorry about that! Turns out my rustfmt was old. Would you prefer me to avoid rustfmt altogether, or add a config like the two linked from serde & rustmft?

Comment (Keats) [inline src/crypto.rs]: Avoid it for now please, we can run it later

File: src/serialization.rs (hunk 2 above)
Comment (mike-engel) [inline src/serialization.rs]: Thanks for the clarification. If it's not obvious, I'm new to Rust, and this result was me fighting with lifetime errors from the compiler 😞

Comment (dtolnay) [inline src/serialization.rs]: I filed https://github.com/Manishearth/rust-clippy/issues/1689 to catch this case with a warning in Clippy. Apparently even the standard library does it wrong sometimes!

**Comment by mike-engel** [comment]
Ok, updated now with `DeserializedOwned`, which made everything much better. Sorry I didn't see that in the docs before, and thanks to @dtolnay for the suggestions, clarifications, work the work on serde 😄 

Tests and benchmarks run fine. If there's anything else you'd like me to fix/edit/add, let me know.

File: src/serialization.rs [hunk 4]
```diff
@@ -10,7 +10,7 @@ use header::Header;
 
 /// The return type of a successful call to decode
 #[derive(Debug)]
-pub struct TokenData<T: Deserialize> {
+pub struct TokenData<T: DeserializeOwned> {
```
Comment (dtolnay) [inline src/serialization.rs]: This bound should be removed: just `pub struct TokenData<T> { ... }`. See https://github.com/Manishearth/rust-clippy/issues/1689.

**Comment by Keats** [comment]
Thanks to both of you!
